README.md
Introduction
============

This utility allows for secure remote tunnel connections to the [Mogotest](http://mogotest.com/) service.  For more
information on how to use this tool, please see the [Mogotest documentation](http://docs.mogotest.com/display/HOWTO/Test+a+Site+on+a+Private+Network).

Building
========

This project will produced a fat JAR including all the necessary dependencies.  In order to build the project you must
have SBT >= 0.12.x installed.  Then you can check out the project and run:

    $ sbt proguard

In the `target/` directory you'll find both the skinny and fat JARs.  The fat one will be named
`mogotest-ssh-tunnel-client-VERSION.min.jar`.

While both JARs will work fine, the fat JAR is preferable in most cases because no classpath management is necessary.
I.e., all you need is this single, executable JAR in order to run the Mogotest tunnel tool.

ChangeLog.md
# 1.2.0 - 2013-03-28

  * Added support for authenticating with a proxy server.
  * Simplified proxy server configuration by synchronizing HTTPS and HTTP proxy info if the latter is not configured.
  * Added "-v" and "--version" flags to print out the version of the tool.
  * Updated Scala 2.9.2 to 2.10.1 to pick up some bug fixes and performance improvements.
  * Updated JCommander from 1.27 to 1.30 to get some bug fixes.
  * Updated JSch from 0.1.48 to 0.1.49 to get some bug fixes and performance improvements.
  * Updated scalaj-http from 0.3.2 to 0.3.6 to get something that will work with Scala 2.10.1.

# 1.1.0 - 2012-10-12

  * Changed the default HTTP tunnel server port from 8888 to 80 to better work out of the box with restrictive networks.
  * Updated Scala 2.9.1 to Scala 2.9.2 to pick up some bug fixes.
  * Updated to SBT 0.12.0 for building the app and updated corresponding SBT plugins.
  * Updated JCommander from 1.18 to 1.27 to get some bug fixes.
  * Updated scalaj-http from 0.2.9 to 0.3.2 to get some bug fixes and Scala 2.9.2 compatibility.
  * Updated JSch from 0.1.44 to 0.1.48 to get some performance enhancements and bug fixes in SSH session handling.

# 1.0.1 - 2012-02-01

  * Increased HTTP timeout values since the defaults were aggressively low.

# 1.0.0 - 2011-09-11

  * Initial release

src/main/scala/com/mogotest/localtunnel/Tunnel.scala
/**
 * Copyright 2011 Mogoterra, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mogotest.localtunnel

import scalaj.http.{Http, HttpOptions}
import com.jcraft.jsch.{JSch, KeyPair}
import java.io.ByteArrayOutputStream

class Tunnel(tunnelHost: String, reflectedHost: String, reflectedPort: Int)
{
  val jsch = new JSch()
  val keyPair = KeyPair.genKeyPair(jsch, KeyPair.RSA, 2048)

  def registerTunnel(): ServerResponse =
  {
    val publicKey = new ByteArrayOutputStream()
    keyPair.writePublicKey(publicKey, "localtunnel")

    val res = new ServerResponse(Http.post(String.format("http://%s/", tunnelHost)).
                                      option(HttpOptions.connTimeout(1000)).
                                      option(HttpOptions.readTimeout(5000)).
                                      params("key" -> publicKey.toString).asString)

    if (res.errorMessage != null)
    {
      println(String.format("  [Error] %s", res.errorMessage))
      sys.exit(-1)
    }

    res
  }

  def startTunnel(tunnelConfig: ServerResponse, mogotest: Mogotest)
  {
    val privateKey = new ByteArrayOutputStream()
    keyPair.writePrivateKey(privateKey)
    jsch.addIdentity(tunnelConfig.user, privateKey.toByteArray, keyPair.getPublicKeyBlob, null)

    val session = jsch.getSession(tunnelConfig.user, tunnelConfig.sshHost)
    session.setConfig("StrictHostKeyChecking", "no")

    session.connect

    if (!session.isConnected)
    {
      println("Invalid SSH key supplied.")
      sys.exit(-1)
    }

    session.setPortForwardingR(tunnelConfig.throughPort, reflectedHost, reflectedPort)

    if (!tunnelConfig.banner.isEmpty)
      println("  " + tunnelConfig.banner)

    println(String.format("   You're good to go. Any tests you run against '%s' on Mogotest will now access the site running on http://%s:%s/.",
      mogotest.testHost, reflectedHost, reflectedPort.toString))

    sys.ShutdownHookThread { session.delPortForwardingR(tunnelConfig.throughPort) }
  }
}

src/main/scala/com/mogotest/localtunnel/Mogotest.scala
/**
 * Copyright 2011 Mogoterra, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mogotest.localtunnel

import scalaj.http.{Http, HttpOptions}

case class Mogotest(apiHost: String, apiKey: String, testHost: String, tunnelUrl: String)
{
  def notifyMogotestOfTunnel()
  {
    val responseCode = Http.post(String.format("https://%s/api/v1/ssh_tunnels", apiHost)).
                            option(HttpOptions.connTimeout(1000)).
                            option(HttpOptions.readTimeout(5000)).
                            params("user_credentials" -> apiKey, "hostname" -> testHost,"tunnel_url" -> tunnelUrl).responseCode

    if (responseCode != 201)
    {
      println("  [Error] Unable to register tunnel location with Mogotest API.  Perhaps your credentials are bad.")
      sys.exit(-1)
    }
  }

  def teardownTunnelInMogotest()
  {
    val responseCode = Http.post(String.format("https://%s/api/v1/ssh_tunnels/destroy", apiHost)).
                            option(HttpOptions.connTimeout(1000)).
                            option(HttpOptions.readTimeout(5000)).
                            params("user_credentials" -> apiKey, "hostname" -> testHost).responseCode

    if (responseCode != 200)
    {
      println("  [Error] Unable to delete tunnel location from Mogotest API.  Perhaps your credentials are bad.")
      sys.exit(-1)
    }
  }
}

src/main/scala/com/mogotest/localtunnel/ServerResponse.scala
/**
 * Copyright 2011 Mogoterra, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mogotest.localtunnel

import scala.util.parsing.json._

class ServerResponse(json: String)
{
  private val decoded = JSON.parseFull(json).get.asInstanceOf[Map[String, Any]]

  val host = decoded.getOrElse("host", null).asInstanceOf[String]
  val throughPort = decoded.getOrElse("through_port", -1).asInstanceOf[Double].toInt
  val banner = decoded.getOrElse("banner", null).asInstanceOf[String]
  val user = decoded.getOrElse("user", null).asInstanceOf[String]
  val errorMessage = decoded.getOrElse("error", null).asInstanceOf[String]

  val sshHost = if (host == null) null else host.split(':').head
}

src/main/scala/com/mogotest/localtunnel/Main.scala
/**
 * Copyright 2011 Mogoterra, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.mogotest.localtunnel

import collection.JavaConversions._
import java.net.{Authenticator, PasswordAuthentication}
import com.beust.jcommander.{ParameterException, JCommander, Parameter}


object Main
{
  object Args
  {
    @Parameter(names = Array("--tunnel_host"), description = "Connect to the named tunnel host (advanced debug mode)", hidden = true)
    var tunnelHost: String = "open.mogotunnel.com:80"

    @Parameter(names = Array("--api_host"), description = "Connect to the named API host for registration (advanced debug mode)", hidden = true)
    var apiHost: String = "mogotest.com"

    @Parameter(names = Array("-h", "--help"), description = "Show this help message.", hidden = true)
    var showHelp: Boolean = false

    @Parameter(names = Array("-v", "--version"), description = "Print out the version.", hidden = true)
    var showVersion: Boolean = false

    @Parameter
    var args: java.util.List[String] = null
  }

  def main(args: Array[String])
  {
    // Scala can't disambiguate method arity when varargs are present.  So, we have to use reflection to find the correct
    // constructor from Java directly.
    val unambiguousConstructor = classOf[JCommander].getConstructors.filter(_.getParameterTypes.length == 1).head
    val parser = unambiguousConstructor.newInstance(Args).asInstanceOf[JCommander]
    parser.setProgramName("java -jar mogotest.jar <mogotest_api_key> <hostname_to_test> <local_port>")

    try
    {
      parser.parse(args.toArray: _*)
    }
    catch
    {
      case e: ParameterException => { println(e.getMessage); parser.usage; sys.exit(-1) }
    }

    if (Args.showHelp)
    {
      parser.usage
      sys.exit(0)
    }

    if (Args.showVersion)
    {
      println("1.2.0")
      sys.exit(0)
    }

    if ((Args.args == null) || (Args.args.length != 3))
    {
      parser.usage
      sys.exit(-1)
    }

    val apiKey = Args.args(0)
    val testHost = Args.args(1)
    val reflectedConnection = Args.args(2)

    val reflectedHost = if (reflectedConnection.contains(':')) reflectedConnection.split(':').head else "127.0.0.1"
    val reflectedPort = if (reflectedConnection.contains(':')) reflectedConnection.split(':').last.toInt else reflectedConnection.toInt

    // Sync HTTPS and HTTP proxy settings if one isn't specified.
    if ((System.getProperty("http.proxyHost") == null) && (System.getProperty("https.proxyHost") != null))
    {
      System.setProperty("http.proxyHost", System.getProperty("https.proxyHost"))
    }

    if ((System.getProperty("http.proxyPort") == null) && (System.getProperty("https.proxyPort") != null))
    {
      System.setProperty("http.proxyPort", System.getProperty("https.proxyPort"))
    }

    val proxyUsername = System.getProperty("https.proxyUsername")
    val proxyPassword = System.getProperty("https.proxyPassword")

    // Set up proxy authentication if credentials were supplied.
    if ((proxyUsername != null) && (proxyPassword != null))
    {
      Authenticator.setDefault(new Authenticator
      {
        override def getPasswordAuthentication : PasswordAuthentication = new PasswordAuthentication(proxyUsername, proxyPassword.toCharArray)
      })
    }

    val tunnel = new Tunnel(Args.tunnelHost, reflectedHost, reflectedPort)
    val response = tunnel.registerTunnel

    val mogotest = new Mogotest(Args.apiHost, apiKey, testHost, response.host)
    mogotest.notifyMogotestOfTunnel

    sys.addShutdownHook { mogotest.teardownTunnelInMogotest }

    try
    {
      tunnel.startTunnel(response, mogotest)
    }
    catch
    {
      case e: Exception => { println(e.getMessage); parser.usage; sys.exit(-1) }
    }

    while (true)
      Thread.sleep(1000)
  }
}